### PR TITLE
build: run integration tests with bazel

### PR DESCRIPTION
**Pull request by gregmagolan**
Integration tests passing locally and on CI.

```
//integration/clover:test                                                PASSED in 84.4s
//integration/common-engine:test                                         PASSED in 147.5s
//integration/express-engine-ivy:test                                    PASSED in 165.1s
//integration/express-engine-ivy-hybrid:test                             PASSED in 172.3s
//modules/builders:unit_test                                             PASSED in 247.9s
//modules/common:unit_test                                               PASSED in 0.8s
//modules/common/schematics:unit_tests                                   PASSED in 3.7s
//modules/express-engine:unit_test                                       PASSED in 1.5s
//modules/express-engine/schematics:unit_tests                           PASSED in 4.0s
```

I haven't removed the legacy integration test job yet in this PR which is still passing as the tests are configured to run both under bazel and outside of bazel. The trouble with adding the integration tests to the existing bazel job, however, is that it might make the feedback for unit test failures much slower for some PRs. Best option might be to run the bazel integration tests in a step after the unit tests pass similar to how the legacy integration test step is run currently.

### CODE REVIEW

File: patches/@angular+dev-infra-private+0.0.0-e5a5a4c9edf515bcafb9b72a90fe2ad3e811df1d.patch [hunk 1]
```diff
@@ -0,0 +1,71 @@
+diff --git a/node_modules/@angular/dev-infra-private/bazel/browsers/browser_archive_repo.bzl b/node_modules/@angular/dev-infra-private/bazel/browsers/browser_archive_repo.bzl
```
Comment (gregmagolan) [inline patches/@angular+dev-infra-private+0.0.0-e5a5a4c9edf515bcafb9b72a90fe2ad3e811df1d.patch:1]: should upstream this patch to dev-infra-private

File: package.json [hunk 2]
```diff
@@ -104,5 +105,8 @@
     "typescript": "4.8.1-rc",
     "xhr2": "^0.2.0",
     "zone.js": "^0.11.0"
+  },
+  "dependencies": {
+    "patch-package": "^6.4.7"
```
Comment (gregmagolan) [inline package.json:110]: can be removed once dev-infra-private patch is upstreamed

File: modules/common/BUILD.bazel [hunk 3]
```diff
@@ -36,6 +36,16 @@ ng_package(
     ],
 )
 
+# work-around for ng_package not providing its output in runfiles
```
Comment (gregmagolan) [inline modules/common/BUILD.bazel:39]: worth fixing in ng_package?

File: integration/common-engine/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,30 @@
+sh_test(
+    name = "test",
+    srcs = ["test.sh"],
+    args = ["$(NODE_PATH)"],
+    data = glob(["**"], exclude = ["BUILD.bazel", "node_modules", "dist"]) + [
+        "@npm//:node_modules_files",
+        "//modules/builders:npm_package",
+        "//modules/common:npm_package_runfiles",
+        # This eager toolchain fetching should be cleaned up in the future
+        "@nodejs_darwin_amd64//:node_files",
```
Comment (gregmagolan) [inline integration/common-engine/BUILD.bazel:10]: should add select statements here so only the toolchains for the host platform are fetched

File: integration/express-engine-ivy-hybrid/BUILD.bazel [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    name = "test",
+    srcs = ["test.sh"],
+    args = ["$(NODE_PATH)"],
+    data = glob(["**"], exclude = ["BUILD.bazel", "node_modules", "dist"]) + [
+        "@npm//:node_modules_files",
+        "//modules/builders:npm_package",
+        "//modules/common:npm_package_runfiles",
+        "//modules/express-engine:npm_package_runfiles",
+        # This eager toolchain fetching should be cleaned up in the future
+        "@nodejs_darwin_amd64//:node_files",
```
Comment (gregmagolan) [inline integration/express-engine-ivy-hybrid/BUILD.bazel:11]: should add select statements here so only the toolchains for the host platform are fetched

File: integration/clover/BUILD.bazel [hunk 6]
```diff
@@ -0,0 +1,17 @@
+sh_test(
+    name = "test",
+    srcs = ["test.sh"],
+    args = ["$(NODE_PATH)"],
+    data = glob(["**"], exclude = ["BUILD.bazel"]) + [
+        "@npm//:node_modules_files",
+        "//modules/builders:npm_package",
+        "//modules/common:npm_package_runfiles",
+        # This eager toolchain fetching should be cleaned up in the future
+        "@nodejs_darwin_amd64//:node_files",
```
Comment (gregmagolan) [inline integration/clover/BUILD.bazel:10]: should add select statements here so only the toolchains for the host platform are fetched

File: integration/express-engine-ivy/BUILD.bazel [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    name = "test",
+    srcs = ["test.sh"],
+    args = ["$(NODE_PATH)"],
+    data = glob(["**"], exclude = ["BUILD.bazel", "node_modules", "dist"]) + [
+        "@npm//:node_modules_files",
+        "//modules/builders:npm_package",
+        "//modules/common:npm_package_runfiles",
+        "//modules/express-engine:npm_package_runfiles",
+        # This eager toolchain fetching should be cleaned up in the future
+        "@nodejs_darwin_amd64//:node_files",
```
Comment (gregmagolan) [inline integration/express-engine-ivy/BUILD.bazel:11]: should add select statements here so only the toolchains for the host platform are fetched

File: integration/clover/test.sh [hunk 8]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+set -o errexit -o nounset
+
+# put node on the path
```
Comment (gregmagolan) [inline integration/clover/test.sh:4]: is it worth consolidating common bits from the 3 test.sh files to a bash helper library?

File: .circleci/config.yml [hunk 9]
```diff
@@ -74,10 +74,11 @@ jobs:
     resource_class: xlarge
     steps:
       - custom_attach_workspace
+      - browser-tools/install-chrome
```
Comment (gregmagolan) [inline .circleci/config.yml:77]: bazel integration use dev-infra-private browser repositories but this install is still needed for shared libs to be installed

File: .circleci/config.yml [hunk 10]
```diff
@@ -74,10 +74,11 @@ jobs:
...
       - copy-bazel-config
+      - run: sudo apt-get update -y && sudo apt-get install -y rsync
```
Comment (gregmagolan) [inline .circleci/config.yml:79]: integration test.sh files use rsync to setup the tmp sandbox directory

File: WORKSPACE [hunk 11]
```diff
@@ -47,4 +47,18 @@ yarn_install(
     # See https://github.com/angular/dev-infra/pull/446#issuecomment-1059820287 for details.
     symlink_node_modules = True,
     yarn_lock = "//:yarn.lock",
+    data = [
+        "//:patches/@angular+dev-infra-private+0.0.0-e5a5a4c9edf515bcafb9b72a90fe2ad3e811df1d.patch",
+    ],
+    manual_build_file_contents = """\
+filegroup(
+    name = "node_modules_files",
```
Comment (gregmagolan) [inline WORKSPACE:55]: used by integration tests

File: patches/@angular+dev-infra-private+0.0.0-e5a5a4c9edf515bcafb9b72a90fe2ad3e811df1d.patch (hunk 1 above)
Comment (alan-agius4) [reply patches/@angular+dev-infra-private+0.0.0-e5a5a4c9edf515bcafb9b72a90fe2ad3e811df1d.patch:1]: Should or will😀?

File: integration/clover/test.sh (hunk 8 above)
Comment (alan-agius4) [reply integration/clover/test.sh:4]: Yes, I think so. Although we can do that as a follow up.

File: patches/@angular+dev-infra-private+0.0.0-e5a5a4c9edf515bcafb9b72a90fe2ad3e811df1d.patch (hunk 1 above)
Comment (gregmagolan) [reply patches/@angular+dev-infra-private+0.0.0-e5a5a4c9edf515bcafb9b72a90fe2ad3e811df1d.patch:1]: Will :) I'll put up a PR

File: package.json (hunk 2 above)
Comment (gregmagolan) [reply package.json]: done; also put up a pre-factor related to updating to the latest dev tooling releases https://github.com/angular/universal/pull/2839

File: patches/@angular+dev-infra-private+0.0.0-e5a5a4c9edf515bcafb9b72a90fe2ad3e811df1d.patch (hunk 1 above)
Comment (gregmagolan) [reply patches/@angular+dev-infra-private+0.0.0-e5a5a4c9edf515bcafb9b72a90fe2ad3e811df1d.patch]: ✅

**Comment by alan-agius4** [comment]
@gregmagolan can you please rebase.

File: integration/express-engine-ivy/BUILD.bazel (hunk 7 above)
Comment (gregmagolan) [reply integration/express-engine-ivy/BUILD.bazel]: done

**Comment by gregmagolan** [comment]
> @gregmagolan can you please rebase.

Done. And cleaned up the node & chromium files so they are selected based on the target platform (assuming the target is the same as execution which should be true since those tests do not run remote)

File: integration/express-engine-ivy-hybrid/BUILD.bazel (hunk 5 above)
Comment (gregmagolan) [reply integration/express-engine-ivy-hybrid/BUILD.bazel]: Done

File: modules/common/BUILD.bazel (hunk 3 above)
Comment (josephperrott) [reply modules/common/BUILD.bazel:39]: Yeah, we definitely should and then we can fixup this later.

**Review by josephperrott — APPROVED** [review]
LGTM

**Pull request merged**